File: threading/python/04_deadlock_prevention.py
"""
QUESTION 4: Deadlock Detection and Prevention (Rubrik/Databricks Critical!)
============================================================================

Problem Statement:
------------------
Given a system where multiple threads acquire multiple locks, implement:
1. A mechanism to detect potential deadlocks
2. A lock ordering strategy to prevent deadlocks
3. A timeout-based approach to avoid deadlocks

Companies: Rubrik (distributed storage), Databricks (distributed computing) - CRITICAL topic!

Key Concepts:
- Coffman conditions for deadlock
- Lock ordering
- Timeout-based acquisition
- Deadlock detection graphs
"""

import threading
import time
from contextlib import contextmanager
from typing import List, Set, Dict
import random


# ============================================================================
# DEADLOCK EXAMPLE - What NOT to do
# ============================================================================

def deadlock_example():
    """
    CLASSIC DEADLOCK SCENARIO:
    Thread 1: Lock A -> Lock B
    Thread 2: Lock B -> Lock A
    
    If Thread 1 holds A and waits for B, while Thread 2 holds B and waits for A,
    neither can proceed = DEADLOCK!
    """
    lock_a = threading.Lock()
    lock_b = threading.Lock()
    
    def thread1():
        with lock_a:
            print("Thread 1 acquired Lock A")
            time.sleep(0.1)  # Increase chance of deadlock
            with lock_b:
                print("Thread 1 acquired Lock B")
    
    def thread2():
        with lock_b:
            print("Thread 2 acquired Lock B")
            time.sleep(0.1)
            with lock_a:
                print("Thread 2 acquired Lock A")
    
    # This WILL deadlock!
    # t1 = threading.Thread(target=thread1)
    # t2 = threading.Thread(target=thread2)
    # t1.start(); t2.start()


# ============================================================================
# SOLUTION 1: Lock Ordering (Most Common Interview Answer)
# ============================================================================

class OrderedLock:
    """
    Lock with ordering to prevent deadlocks.
    
    EXPLANATION:
    ------------
    Assign a unique ID to each lock. Always acquire locks in ID order.
    This breaks the circular wait condition (one of the 4 Coffman conditions).
    
    COFFMAN CONDITIONS (all 4 needed for deadlock):
    1. Mutual Exclusion: Resources cannot be shared
    2. Hold and Wait: Thread holds resources while waiting for others
    3. No Preemption: Resources cannot be forcibly taken
    4. Circular Wait: Circular chain of threads waiting for each other
    
    Breaking ANY ONE condition prevents deadlock!
    """
    
    _id_counter = 0
    _id_lock = threading.Lock()
    
    def __init__(self, name: str = ""):
        with OrderedLock._id_lock:
            self._id = OrderedLock._id_counter
            OrderedLock._id_counter += 1
        self._lock = threading.Lock()
        self.name = name or f"Lock-{self._id}"
    
    @property
    def id(self):
        return self._id
    
    def acquire(self):
        self._lock.acquire()
    
    def release(self):
        self._lock.release()
    
    def __enter__(self):
        self.acquire()
        return self
    
    def __exit__(self, *args):
        self.release()


@contextmanager
def acquire_locks_ordered(*locks: OrderedLock):
    """
    Acquire multiple locks in consistent order to prevent deadlock.
    """
    # Sort locks by ID
    sorted_locks = sorted(locks, key=lambda x: x.id)
    
    acquired = []
    try:
        for lock in sorted_locks:
            lock.acquire()
            acquired.append(lock)
            print(f"Acquired {lock.name}")
        yield
    finally:
        # Release in reverse order
        for lock in reversed(acquired):
            lock.release()
            print(f"Released {lock.name}")


# ============================================================================
# SOLUTION 2: Timeout-Based Lock Acquisition
# ============================================================================

class TimeoutLock:
    """
    Lock with timeout to avoid deadlocks.
    
    EXPLANATION:
    ------------
    If we can't acquire a lock within timeout, back off and retry.
    This breaks the "hold and wait" condition.
    """
    
    def __init__(self, name: str = ""):
        self._lock = threading.Lock()
        self.name = name
    
    def acquire_with_timeout(self, timeout: float = 1.0) -> bool:
        """Try to acquire lock with timeout. Returns True if acquired."""
        return self._lock.acquire(timeout=timeout)
    
    def release(self):
        self._lock.release()


def acquire_multiple_with_timeout(locks: List[TimeoutLock], timeout: float = 1.0) -> bool:
    """
    Try to acquire all locks. If any fails, release all and return False.
    """
    acquired = []
    try:
        for lock in locks:
            if lock.acquire_with_timeout(timeout):
                acquired.append(lock)
            else:
                # Failed to acquire - release all and return
                for l in acquired:
                    l.release()
                return False
        return True
    except Exception:
        for l in acquired:
            l.release()
        raise


def demo_ordered_locking():
    """Demonstrate deadlock-free locking with ordering"""
    lock_a = OrderedLock("A")
    lock_b = OrderedLock("B")
    
    def thread1():
        with acquire_locks_ordered(lock_a, lock_b):
            print("Thread 1 working with both locks")
            time.sleep(0.1)
    
    def thread2():
        # Even though we pass B first, they'll be acquired in order
        with acquire_locks_ordered(lock_b, lock_a):
            print("Thread 2 working with both locks")
            time.sleep(0.1)
    
    t1 = threading.Thread(target=thread1)
    t2 = threading.Thread(target=thread2)
    t1.start(); t2.start()
    t1.join(); t2.join()
    print("No deadlock!")


if __name__ == "__main__":
    demo_ordered_locking()

File: threading/cpp/04_deadlock_prevention.cpp
/*
 * QUESTION 4: Deadlock Detection and Prevention (Rubrik/Databricks Critical!)
 * ============================================================================
 * 
 * Problem: Implement mechanisms to detect and prevent deadlocks.
 * 
 * Key Concepts: std::lock, std::scoped_lock, lock ordering, try_lock
 * 
 * COFFMAN CONDITIONS (all 4 needed for deadlock):
 * 1. Mutual Exclusion   2. Hold and Wait
 * 3. No Preemption      4. Circular Wait
 * 
 * Compile: g++ -std=c++17 -pthread 04_deadlock_prevention.cpp -o deadlock
 */

#include <iostream>
#include <thread>
#include <mutex>
#include <chrono>
#include <vector>

// SOLUTION 1: std::scoped_lock (C++17) - Deadlock-free multiple lock acquisition
class Account {
public:
    std::mutex mutex;
    int balance;
    int id;
    
    Account(int id, int balance) : id(id), balance(balance) {}
};

void transfer_safe(Account& from, Account& to, int amount) {
    /*
     * EXPLANATION:
     * std::scoped_lock acquires multiple locks in deadlock-free manner.
     * Uses deadlock avoidance algorithm internally (try-and-back-off).
     * This is the RECOMMENDED approach in modern C++!
     */
    std::scoped_lock lock(from.mutex, to.mutex);
    
    if (from.balance >= amount) {
        from.balance -= amount;
        to.balance += amount;
        std::cout << "Transferred " << amount << " from " << from.id 
                  << " to " << to.id << std::endl;
    }
}

// SOLUTION 2: std::lock + std::adopt_lock (C++11)
void transfer_safe_cpp11(Account& from, Account& to, int amount) {
    /*
     * EXPLANATION:
     * std::lock() locks multiple mutexes without deadlock.
     * std::adopt_lock tells lock_guard the mutex is already locked.
     */
    std::lock(from.mutex, to.mutex);
    std::lock_guard<std::mutex> lock1(from.mutex, std::adopt_lock);
    std::lock_guard<std::mutex> lock2(to.mutex, std::adopt_lock);
    
    if (from.balance >= amount) {
        from.balance -= amount;
        to.balance += amount;
    }
}

// SOLUTION 3: Lock Ordering (Manual approach)
void transfer_ordered(Account& from, Account& to, int amount) {
    /*
     * EXPLANATION:
     * Always acquire locks in consistent order (by ID).
     * Breaks CIRCULAR WAIT condition -> No deadlock!
     */
    Account* first = &from;
    Account* second = &to;
    
    if (from.id > to.id) {
        std::swap(first, second);
    }
    
    std::lock_guard<std::mutex> lock1(first->mutex);
    std::lock_guard<std::mutex> lock2(second->mutex);
    
    if (from.balance >= amount) {
        from.balance -= amount;
        to.balance += amount;
    }
}

// SOLUTION 4: try_lock with timeout (Avoiding indefinite wait)
void transfer_try(Account& from, Account& to, int amount) {
    /*
     * EXPLANATION:
     * Try to acquire locks. If fail, release and retry.
     * Breaks HOLD AND WAIT condition.
     */
    while (true) {
        if (from.mutex.try_lock()) {
            if (to.mutex.try_lock()) {
                // Got both locks
                if (from.balance >= amount) {
                    from.balance -= amount;
                    to.balance += amount;
                }
                to.mutex.unlock();
                from.mutex.unlock();
                return;
            }
            from.mutex.unlock();  // Release first lock
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(1));  // Backoff
    }
}

int main() {
    Account acc1(1, 1000);
    Account acc2(2, 1000);
    
    std::vector<std::thread> threads;
    
    // Multiple threads transferring in both directions
    for (int i = 0; i < 5; ++i) {
        threads.emplace_back(transfer_safe, std::ref(acc1), std::ref(acc2), 100);
        threads.emplace_back(transfer_safe, std::ref(acc2), std::ref(acc1), 50);
    }
    
    for (auto& t : threads) t.join();
    
    std::cout << "Account 1: " << acc1.balance << std::endl;
    std::cout << "Account 2: " << acc2.balance << std::endl;
    std::cout << "Total: " << acc1.balance + acc2.balance << " (should be 2000)" << std::endl;
    
    return 0;
}

